compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -880,6 +880,97 @@ impl<'a> Parser<'a> {
         }
     }
 
+    pub(super) fn check_anonymous_enum(&mut self, span: Span, tys: &[P<Ty>]) -> PResult<'a, ()> {
+        use std::fmt::Write;
+        if tys.len() <= 1 {
+            return Ok(());
+        }
+
+        fn variant_name_and_ty<'a>(
+            ty: &'a Ty,
+            lifetimes: &mut FxHashSet<&'a str>,
+        ) -> Option<(String, String)> {
+            match &ty.kind {
+                TyKind::Path(_, path) => {
+                    let mut name = String::new();
+                    let mut ty_string = String::new();
+
+                    if let Some(seg) = path.segments.iter().last() {
+                        name.push_str(seg.ident.name.as_str());
+                        ty_string.push_str(seg.ident.name.as_str());
+
+                        if let Some(_args) = &seg.args {
+                            ty_string.push('<');
+                            ty_string.push_str("...");
+                            ty_string.push('>');
+                        }
+                    }
+
+                    Some((name, ty_string))
+                }
+                TyKind::Rptr(lifetime, ty) => {
+                    if let Some((mut name, ty)) = variant_name_and_ty(&ty.ty, lifetimes) {
+                        name.push_str("Ref");
+                        let lifetime =
+                            lifetime.as_ref().map(|l| l.ident.name.as_str()).unwrap_or("'lifetime");
+                        lifetimes.insert(lifetime);
+                        Some((name, format!("&{} {}", lifetime, ty)))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
```

Comment:
This wouldn't work for something like `-> [i32; 10] | Vec<i32>`, right?

If you don't want to support every `TyKind`, you can at least provide a `Variant{N}` `name` so that the suggested `enum` makes syntactic sense.